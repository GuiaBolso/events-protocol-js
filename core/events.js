"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseEvent = void 0;
var errors_1 = require("./errors");
var ResponseEvent = /** @class */ (function() {
    function ResponseEvent(event) {
        this.name = event.name;
        this.version = event.version;
        this.id = event.id;
        this.flowId = event.flowId;
        this.payload = event.payload;
        this.identity = event.identity;
        this.auth = event.auth;
        this.metadata = event.metadata;
    }
    ResponseEvent.prototype.isSuccess = function() {
        return this.name.endsWith(":response");
    };
    ResponseEvent.prototype.isError = function() {
        return !this.isSuccess();
    };
    ResponseEvent.prototype.getErrorType = function() {
        if (this.isSuccess()) throw Error("This is not an error event.");
        return errors_1.getErrorType(
            this.name.substring(this.name.lastIndexOf(":"))
        );
    };
    return ResponseEvent;
})();
exports.ResponseEvent = ResponseEvent;
//# sourceMappingURL=events.js.map
