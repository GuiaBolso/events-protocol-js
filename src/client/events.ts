import { getUUID, getSessionUUID } from "../utils/uuid";

export const SUCCESS = 2;

export class Event {
  name: string;

  version: number;

  payload: any | null;

  metadata: Metadata | null;

  auth: Auth | null;

  identity: any | undefined | null;

  flowId: string;

  id: string;
}

type Auth = {
  token: string | null;
  "x-sid": string | null;
  "x-tid": string | null;
};
type Metadata = {
  createdAt: Date;
  origin: string;
};

type HTTPError = {
  message: string;
  code: number;
  flowId: string;
  id: string;
};

const UNAUTHORIZED_REQUEST = "others";
const TOKEN_KEY = "sessionToken";

const DEFAULT_ORIGIN = "web";

/**
 * https://github.com/GuiaBolso/events-protocol
 */
const EVENT_SUCCESS = "response";
const REDIRECT = "redirect";

export class AuthError extends Error {
  flowId: string;

  id: string;

  constructor(message: string, flowId: string, id: string) {
    super(message);
    this.flowId = flowId;
    this.id = id;
  }
}

export class HttpError extends Error {
  code: number;

  response: Response;

  flowId: string;

  id: string;

  constructor(
    message: string,
    code: number,
    response: Response,
    flowId: string,
    id: string
  ) {
    super(message);
    this.code = code;
    this.response = response;
    this.flowId = flowId;
    this.id = id;
  }
}

export class EventError extends Error {
  code: string;

  payload: Record<string, any>;

  flowId: string;

  id: string;

  constructor(
    message: string,
    code: string,
    payload: Record<string, any>,
    flowId: string,
    id: string
  ) {
    super(message);
    this.code = code;
    this.payload = payload;
    this.flowId = flowId;
    this.id = id;
  }
}

export function handleHttpResponse(
  response: Response,
  flowId: string,
  id: string
): Response {
  const baseStatus = Math.floor(response.status / 100);

  if (baseStatus === SUCCESS) {
    return response;
  }

  throw new HttpError(
    response.statusText,
    response.status,
    response,
    flowId,
    id
  );
}

export function handleJSON(response: Response): Promise<Response> {
  return response.json();
}

/**
 *  get latest chunk of event name. Eg.: 'event:name:response' -> 'response'
 */
export const getEventChunks = (eventName: string): Array<string> =>
  eventName ? eventName.split(":") : [];
export const last = (arr: Array<string>): string =>
  arr.length ? arr[arr.length - 1] : "";
export const getEventLastChunk = (eventName: string): string =>
  last(getEventChunks(eventName));

/**
 *  get request event name and split into name and version. Eg.: 'event:name:v2' -> { name: 'event:name', version: 2 }
 */
const parseEventName = (eventName: string): any => {
  const [, name, , version]: Array<string> =
    eventName.match(/(.+?)(:v(\d+))?$/i) || [];
  return {
    name,
    version
  };
};

export function handleEventResponse(
  eventResponse: Event,
  nameHandler: Function = getEventLastChunk
): Promise<Event> {
  const baseStatus = nameHandler(eventResponse.name);

  if (baseStatus === EVENT_SUCCESS || baseStatus === REDIRECT) {
    return Promise.resolve(eventResponse);
  }

  const errorMessage =
    eventResponse.payload && eventResponse.payload.message
      ? eventResponse.payload.message
      : baseStatus;
  throw new EventError(
    errorMessage,
    baseStatus,
    eventResponse.payload || {},
    eventResponse.flowId,
    eventResponse.id
  );
}

export const handleAuth = (tokenKey: string) => (
  isAuthorized: boolean,
  event: Event
): void => {
  if (isAuthorized) {
    if (!event.auth) {
      throw new AuthError(
        "Expected auth, but none found",
        event.flowId,
        event.id
      );
    } else if (!event.auth.token) {
      throw new AuthError(
        `Expected ${tokenKey} in auth, but none found`,
        event.flowId,
        event.id
      );
    }
  }
};

export const createEvent = (
  { name, version, payload, auth, metadata }: Event,
  config: any,
  baseConfig: any = {
    uuidResolver: getUUID,
    localUuidResolver: getSessionUUID,
    dateResolver: (): Date => new Date(),
    ...config
  }
): Event => {
  if (!metadata || !metadata.origin || metadata.origin === DEFAULT_ORIGIN) {
    console.info("Your application should use its name as metadata.origin");
  }
  return {
    name,
    version,
    payload,
    flowId: baseConfig.uuidResolver(),
    id: baseConfig.uuidResolver(),
    auth: {
      token: null,
      ...auth,
      "x-sid": baseConfig.localUuidResolver(),
      "x-tid": baseConfig.uuidResolver()
    },
    metadata: {
      ...metadata,
      origin: DEFAULT_ORIGIN,
      createdAt: baseConfig.dateResolver()
    },
    identity: {}
  };
};

export const intoEvent = (json: any): Event => ({
  name: json.name,
  version: json.version,
  payload: json.payload,
  metadata: json.metadata,
  auth: json.auth,
  flowId: json.flowId,
  id: json.id,
  identity: {}
});

export const prepareURL = (
  hostname: string,
  noauthURL: string,
  isAuthorized = true
): string =>
  isAuthorized ? `${hostname}/events/` : `${hostname}/events/${noauthURL}`;

export const generateFetchEvent = (config: any = {}): Function => (
  event: Event,
  isAuthorized: boolean | undefined = true
): Promise<Event> => {
  const {
    hostname = "",
    noauthURL = `${UNAUTHORIZED_REQUEST}`,
    fetchResolver = fetch, //eslint-disable-line
    tokenKey = TOKEN_KEY,
    authHandler = handleAuth(tokenKey),
    httpResponseHandler = handleHttpResponse,
    jsonParser = handleJSON,
    eventResponseHandler = handleEventResponse,
    urlHandler = prepareURL,
    convertIntoEvent = intoEvent
  } = config || {};

  const url = urlHandler(hostname, noauthURL, isAuthorized);

  authHandler(isAuthorized, event);

  return fetchResolver(url, {
    method: "POST",
    headers: {
      "Content-type": "application/json"
    },
    body: JSON.stringify(event)
  })
    .then((response: any) =>
      httpResponseHandler(response, event.flowId || "flowId", event.id || "id")
    )
    .then(jsonParser)
    .then(convertIntoEvent)
    .then(eventResponseHandler);
};

export const generateFetchEventByName = (
  config: any,
  baseConfig: any = {
    eventCreator: createEvent,
    fetchEventGenerator: generateFetchEvent,
    eventNameParser: parseEventName,
    ...config
  }
): Function => (eventName: string, payload: any, confs: any = {}): any => {
  const conf = {
    metadata: {
      origin: DEFAULT_ORIGIN
    },
    identity: {},
    isAuthorized: true,
    auth: {},
    parsedEvent: baseConfig.eventNameParser(eventName),
    ...confs
  };

  const event = baseConfig.eventCreator({
    name: conf.parsedEvent.name,
    version: conf.parsedEvent.version,
    payload,
    auth: conf.auth,
    metadata: baseConfig.metadata || conf.metadata,
    identity: baseConfig.identity || conf.identity
  });

  return baseConfig.fetchEventGenerator(baseConfig)(event, conf.isAuthorized);
};
