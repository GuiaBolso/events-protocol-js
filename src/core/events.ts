import { EventErrorType, getErrorType } from "core/errors";

export interface Event {
    name: string;
    version: number;
    payload: any;
    metadata: Metadata;
    auth: Auth;
    identity: any;
    flowId: string;
    id: string;
}

interface Auth {
    token?: string | null;
    "x-sid"?: string;
    "x-tid"?: string;
}

interface Metadata {
    createdAt?: Date;
    origin?: string;
}

export class ResponseEvent implements Event {
    name: string;
    version: number;
    payload: any;
    metadata: Metadata;
    auth: Auth;
    identity: any;
    flowId: string;
    id: string;

    constructor(event: Event) {
        this.name = event.name;
        this.version = event.version;
        this.id = event.id;
        this.flowId = event.flowId;
        this.payload = event.payload;
        this.identity = event.identity;
        this.auth = event.auth;
        this.metadata = event.metadata;
    }

    isSuccess(): boolean {
        return this.name.endsWith(":response");
    }

    isError(): boolean {
        return !this.isSuccess();
    }

    getErrorType(): EventErrorType {
        if (this.isSuccess()) throw Error("This is not an error event.");
        return getErrorType(this.name.substring(this.name.lastIndexOf(":")));
    }
}
