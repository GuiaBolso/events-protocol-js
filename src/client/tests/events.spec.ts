import * as events from "../events";

const generateMockEvent = (): events.Event => ({
  flowId: "",
  id: "",
  payload: {},
  identity: {},
  auth: {
    "x-sid": "",
    "x-tid": "",
    token: ""
  },
  metadata: {
    origin: "test",
    createdAt: new Date()
  },
  version: 0,
  name: ""
});

describe("handleEventResponse", () => {
  it("must return response when name ends with `response`", async () => {
    const event = { ...generateMockEvent(), name: `event:name:response` };

    expect(await events.handleEventResponse(event)).toBe(event);
  });

  it("must throws Event error with message when name`s end is different from response and payload exists", async () => {
    const event = {
      ...generateMockEvent(),
      name: `event:name:unauthorized`,
      payload: {
        message: "User is not authorized"
      }
    };

    try {
      await events.handleEventResponse(event);
      throw new Error("");
    } catch (e) {
      expect(e.message).toBe("User is not authorized");
    }
  });

  it("must throws Event error with baseStatus when name`s end is different from response and payload does not exist", async () => {
    const event = {
      ...generateMockEvent(),
      name: `event:name:unauthorized`
    };

    try {
      await events.handleEventResponse(event);
      throw new Error("no error thrown");
    } catch (e) {
      expect(e.message).toBe("unauthorized");
    }
  });
});

describe("createEvent", () => {
  it("must convert parameters into a serializable event", () => {
    const e = events.createEvent(
      {
        ...generateMockEvent(),
        name: "event:name",
        version: 1,
        payload: {
          myProperty: 10
        },
        metadata: {
          origin: "test",
          createdAt: new Date()
        },
        auth: {
          "x-sid": "",
          "x-tid": "",
          token: "fa3w43a4-a342f23r-a34ra34234-3rqrqwfwdf"
        }
      },
      {
        uuidResolver: () => 0,
        localUuidResolver: () => 0,
        dateResolver: () => 0
      }
    );
    expect(JSON.stringify(e)).toMatchSnapshot();
  });
});

describe("getEventLastChunk", () => {
  it("must return the last chunk of a common event name", () => {
    expect(events.getEventLastChunk("event:name:response")).toBe("response");
  });
  it("must return the name of event if it is a simple name", () => {
    expect(events.getEventLastChunk("response")).toBe("response");
  });
  it("must return an empty string if it is no name", () => {
    expect(events.getEventLastChunk("")).toBe("");
  });
});

describe("handleAuth", () => {
  it("must throw no errors if it is authorized and auth is present in event", () => {
    expect(
      events.handleAuth("token")(true, {
        ...generateMockEvent(),
        name: "event:name",
        version: 1,
        auth: {
          token: "token",
          "x-sid": "",
          "x-tid": ""
        }
      })
    ).toBe(undefined);
  });

  it("must throw errors if it is authorized and auth is not present in event", () => {
    expect(() =>
      events.handleAuth("token")(true, {
        ...generateMockEvent(),
        auth: null,
        name: "event:name",
        version: 1
      })
    ).toThrow("Expected auth, but none found");
  });

  it("must throw errors if it is authorized and token is not present in auth, in event", () => {
    expect(() =>
      events.handleAuth("token")(true, {
        ...generateMockEvent(),
        name: "event:name",
        version: 1
      })
    ).toThrow("Expected token in auth, but none found");
  });

  it("must throw no errors if it is not authorized and no auth is present", () => {
    expect(
      events.handleAuth("token")(false, {
        ...generateMockEvent(),
        name: "event:name",
        version: 1
      })
    ).toBe(undefined);
  });

  it("must throw no errors if it is not authorized and auth is present", () => {
    expect(
      events.handleAuth("token")(false, {
        ...generateMockEvent(),
        name: "event:name",
        version: 1,
        auth: {
          token: "token",
          "x-sid": "",
          "x-tid": ""
        }
      })
    ).toBe(undefined);
  });
});

describe("generateFetchEvent", () => {
  const fetchResolver = jest.fn(() => Promise.resolve(10));
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const authHandler = jest.fn(() => {});
  const httpResponseHandler = jest.fn(() => Promise.resolve(20));
  const jsonParser = jest.fn(() => Promise.resolve(30));
  const convertIntoEvent = jest.fn(() => Promise.resolve(40));
  const eventResponseHandler = jest.fn(() => Promise.resolve(50));
  const urlHandler = jest.fn(() => "url");

  const config = {
    hostname: "hostname",
    noauthURL: `unauthorized`,
    fetchResolver,
    tokenKey: "token",
    authHandler,
    httpResponseHandler,
    jsonParser,
    convertIntoEvent,
    eventResponseHandler,
    urlHandler
  };

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("must resolve and pipe actions to fetch event", async () => {
    const event = {
      name: "event:name",
      version: 1,
      payload: {
        myProperty: 10
      },
      metadata: {
        code: 3
      },
      auth: {
        token: "fa3w43a4-a342f23r-a34ra34234-3rqrqwfwdf"
      }
    };

    const result = await events.generateFetchEvent(config)(event, true, "1");

    expect(urlHandler).toHaveBeenCalledWith("hostname", `unauthorized`, true);
    expect(authHandler).toHaveBeenCalledWith(true, event);
    expect(fetchResolver).toHaveBeenCalled();
    expect(httpResponseHandler).toHaveBeenCalledWith(10, "flowId", "id");
    expect(jsonParser).toHaveBeenCalledWith(20);
    expect(convertIntoEvent).toHaveBeenCalledWith(30);
    expect(eventResponseHandler).toHaveBeenCalledWith(40);
    expect(result).toBe(50);
  });
});

describe("generateFetchEventByName", () => {
  const event = {};
  const fetchEventGenerator = (): jest.Mock<Promise<number>> =>
    jest.fn(async () => 10);
  const eventCreator = jest.fn(() => event);
  const eventNameParser = (): events.Event => ({
    ...generateMockEvent(),
    name: "event:name",
    version: 2
  });
  const payload = {};
  const auth = {};

  const config = {
    fetchEventGenerator,
    eventCreator,
    eventNameParser
  };

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("must resolve name and version and pipe actions to fetch event", async () => {
    const result = await events.generateFetchEventByName(config)(
      "event:name:v2",
      payload,
      {
        isAuthorized: true,
        auth
      }
    );

    expect(eventCreator).toHaveBeenCalledWith({
      name: "event:name",
      identity: {},
      metadata: { origin: "web" },
      version: 2,
      payload,
      auth
    });

    expect(result).not.toBe(null);
  });
});
