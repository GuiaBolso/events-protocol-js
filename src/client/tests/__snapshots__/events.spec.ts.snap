// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createEvent must convert parameters into a serializable event 1`] = `"{\\"name\\":\\"event:name\\",\\"version\\":1,\\"payload\\":{\\"myProperty\\":10},\\"flowId\\":0,\\"id\\":0,\\"auth\\":{\\"x-sid\\":0,\\"x-tid\\":0,\\"token\\":\\"fa3w43a4-a342f23r-a34ra34234-3rqrqwfwdf\\"},\\"metadata\\":{\\"origin\\":null,\\"createdAt\\":0},\\"identity\\":{}}"`;
